src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectIdentityHashCodeTracker.kt

```diff
@@ -0,0 +1,110 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2024 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck.strategy.managed
+
+import org.jetbrains.kotlinx.lincheck.isInTraceDebuggerMode
+import org.jetbrains.kotlinx.lincheck.util.UnsafeHolder
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
+
+
+private typealias Id = Long
+
+/**
+ * This class stores initial identity hashcodes of created objects.
+ * When the program is rerun the firstly calculated hashcodes are left as is and used to
+ * substitute new hashcodes via [sun.misc.Unsafe] in the object headers.
+ *
+ * To guarantee correct work, ensure that replays are deterministic.
+ */
+internal class ObjectIdentityHashCodeTracker {
+    companion object {
+        /**
+         * Offset (in bytes) of identity hashcode in an object header.
+         *
+         * @see <a href="https://shipilev.net/jvm/anatomy-quarks/26-identity-hash-code/#_hashcode_storage">JVM Anatomy Quark #26: Identity Hash Code</a>
+         */
+        private const val IDENTITY_HASHCODE_OFFSET: Long = 1L
+    }
+    private val initialHashCodes = ConcurrentHashMap<Id, Int>()
+    private val nextObjectId = AtomicLong(0)
+    private val objectIdAdvances = ConcurrentHashMap<Id, Id>()
+
+    /**
+     * This method substitutes identity hash code of the object in its header with the initial (from the first test execution) identity hashcode.
```

Comment:
Now, this technique is applied to both Lincheck and Trace Debugger. Please do a pass over the documentation/comments in this PR and modify them accordingly where necessary.

Reply:
> Now, this technique is applied to both Lincheck and Trace Debugger.
That's false.

Reply:
I mean, we always put the identity hash code value directly into the object now.

Reply:
Still, this method injects the identity hash code in both cases.